### PR TITLE
RSDK-6054 experiment with using multiple seed values for ik solving in ptgs

### PR DESCRIPTION
**Pull request by biotinker**
Generally speaking, when running gradient descent , starting with a better-scoring value will tend to produce a better result, and faster, than using a poor seed.

This updates the seed value for uncapped length PTGs to be the distance to the intended goal, thus choosing a value close to where they will ultimately wind up.

As part of doing this we also separate out the `DoF` function from `Frame` to be its own interface.

### CODE REVIEW

File: motionplan/tpspace/ptgIK.go [hunk 1]
```diff
@@ -100,32 +100,29 @@ func NewPTGIK(simPTG PTG, logger logging.Logger, refDistLong, refDistShort float
...
-	err := ptg.gridSim.Solve(ctx, internalSolutionGen, seed, solveMetric, nloptSeed)
-	if err != nil {
-		return err
-	}
-	select {
-	case gridSolved = <-internalSolutionGen:
-	default:
+	if seedVals != nil {
+		// use the passed seed as the first dist if in range
+		if seedVals[0].Value < ptg.ptgFrame.DoF()[1].Max {
```
Comment (raybjork) [inline motionplan/tpspace/ptgIK.go:118]: I'd prefer to move this to the caller code and simply use the seedVal as passed in

File: motionplan/tpspace/ptgIK.go [hunk 2]
```diff
@@ -100,32 +100,29 @@ func NewPTGIK(simPTG PTG, logger logging.Logger, refDistLong, refDistShort float
...
 ) error {
 	internalSolutionGen := make(chan *ik.Solution, 1)
 	defer close(internalSolutionGen)
 	var solved *ik.Solution
-	var gridSolved *ik.Solution
 
-	if seed == nil {
-		seed = ptg.defaultSeed
+	seed := make([]referenceframe.Input, 0, len(ptg.defaultSeed))
+	for _, defaultSeedVal := range ptg.defaultSeed {
```
Comment (raybjork) [inline motionplan/tpspace/ptgIK.go:112]: It seems unnecessary to deep copy this

**Review by raybjork — CHANGES_REQUESTED** [review]

File: motionplan/tpspace/ptgIK.go (hunk 2 above)
Comment (biotinker) [reply motionplan/tpspace/ptgIK.go:112]: It is necessary; this function can be run in parallel on this struct, and if we don't deep copy then different threads will be editing one another's data structures.

File: motionplan/tpspace/ptgIK.go (hunk 1 above)
Comment (biotinker) [reply motionplan/tpspace/ptgIK.go]: Done

File: motionplan/tpSpaceRRT.go [hunk 3]
```diff
@@ -465,8 +465,15 @@ func (mp *tpSpaceRRTMotionPlanner) getExtensionCandidate(
 	var targetFunc ik.StateMetric
 	relPose := spatialmath.PoseBetween(nearest.Pose(), randPosNode.Pose())
 	targetFunc = mp.algOpts.goalMetricConstructor(relPose)
+	seedDist := relPose.Point().Norm()
+	seed := tpspace.PTGIKSeed(curPtg)
+	dof := curPtg.DoF()
+	if seedDist < dof[1].Max {
```
Comment (raybjork) [inline motionplan/tpSpaceRRT.go:471]: We're repeating ourselves a little with this code block which is repeated on line 810.  Can we change the PTGIKSeed function to operate on start and goal poses or something similar?  This is pretty minor though and I don't feel strongly if you'd rather preserve the function signature as it currently exists

Comment (biotinker) [reply motionplan/tpSpaceRRT.go:471]: This will go away imminently with Nick S's work on RSDK-6052, which is why I felt okay duplicating this

Comment (biotinker) [reply motionplan/tpSpaceRRT.go:471]: So my inclination is to let it be

**Pull request merged**